### PR TITLE
Sort queries to reduce data divergence

### PR DESCRIPTION
**Pull request by dalg24**
An attempt to take advantage of the fact that if objects are nearby, chances are that they will yield almost the same traversals.

### CODE REVIEW

**Comment by dalg24** [comment]
cuda (P100)  | e9a2bde  | 75ef59b  | 4f133ce7
------------ | -------- | -------- | --------
construction | 0.319332 | 0.312941 | 0.308305
kNN          | 0.394494 | 0.40688  | 0.382948
radius       | 0.264993 | 0.315028 | 0.348413

openmp       | e9a2bde  | 75ef59b  | 4f133ce7
------------ | -------- | -------- | --------
construction | 0.348495 | 0.348303 | 0.350486
kNN          | 0.366683 | 0.368394 | 0.583663
radius       | 0.380238 | 0.38536  | 0.86287

serial       | e9a2bde  | 75ef59b  | 4f133ce7
------------ | -------- | -------- | --------
construction | 3.56007  | 3.50881  | 3.52007
kNN          | 4.29842  | 4.27737  | 6.61133
radius       | 4.51795  | 4.52966  | 9.11296

cuda (K80)   | e9a2bde  | 75ef59b  | 4f133ce7
------------ | -------- | -------- | --------
construction | 1.21553  | 1.21178  | 1.21017
kNN          | 2.07129  | 2.11973  | 2.16152
radius       | 0.962253 | 1.01692  | 1.69827

**Comment by dalg24** [comment]
Timings above were obtained with `--values=10000000 --queries=1000000`.  Note that as the size of the tree and the number of queries decrease, sorting the queries gets less and less advantageous and it eventually penalize us for small problems.

**Comment by dalg24** [comment]
For the record, I did try to also add a level of indirection when accessing queries instead of actually applying the permutation but I found it was actually a bit slower.

File: packages/Search/src/details/DTK_DetailsBatchedQueries.hpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    permuteIndices( Kokkos::View<int const *, DeviceType> permute,
+                    Kokkos::View<T const *, DeviceType> indices,
+                    Kokkos::View<int const *, DeviceType> offset,
+                    Kokkos::View<int const *, DeviceType> tmp_offset )
+    {
+        auto const n = permute.extent( 0 );
+
+        DTK_REQUIRE( offset.extent( 0 ) == n + 1 );
+        DTK_REQUIRE( tmp_offset.extent( 0 ) == n + 1 );
+        DTK_REQUIRE( lastElement( offset ) == indices.extent_int( 0 ) );
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsBatchedQueries.hpp]: Be consistent (at least inside one function), either use `extent` or `extent_int` but don't mix them

File: packages/Search/test/tstDetailsBatchedQueries.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    TEST_THROW( DataTransferKit::Details::BatchedQueries<
+                    DeviceType>::reversePermutation( permute, offset, indices ),
+                DataTransferKit::DataTransferKitException );
+
+    Kokkos::resize( offset, 1 );
+    TEST_NOTHROW( DataTransferKit::Details::BatchedQueries<
+                  DeviceType>::reversePermutation( permute, offset, indices ) );
+
+    DataTransferKit::Details::BatchedQueries<DeviceType>::permuteOffset(
```
Comment (Rombur) [inline packages/Search/test/tstDetailsBatchedQueries.cpp]: Is this a test? If yes either comment or use `TEST_NOTHROW`

Comment (dalg24) [reply packages/Search/test/tstDetailsBatchedQueries.cpp]: I think I just forgot to remove that line.

File: packages/Search/src/details/DTK_DetailsBatchedQueries.hpp (hunk 1 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsBatchedQueries.hpp]: Ok, will fix.

**Comment by dalg24** [comment]
OpenMP runs were obtained with
```
OMP_PROC_BIND=spread
OMP_NUM_THREADS=16
OMP_PLACES=threads
```

**Pull request merged**